File: 0260-single-number-iii/0260-single-number-iii.py
class Solution(object):
    def singleNumber(self, nums):
        _xor = 0
        for num in nums:
            _xor ^= num

        right_bit = _xor & (- _xor) 
        x, y = 0, 0

        for num in nums:
            if (right_bit & num) == 0:
                x ^= num
            else:
                y ^= num
        
        return [x, y]

        

File: 0260-single-number-iii/0260-single-number-iii.java
class Solution {
    public int[] singleNumber(int[] nums) {
        int n = nums.length;
        long _xor = 0;
        for(int i = 0; i<n; i++) _xor ^= nums[i];

        long rightBit = _xor & (-_xor);
        int x = 0, y = 0;

        for(int i = 0; i<n; i++) {
            if((rightBit & nums[i]) == 0) x ^= nums[i];
            else y ^= nums[i];
        }

        return new int[]{x, y};
    }
}